### PR TITLE
AWS AL2023 support

### PR DESCRIPTION
**Pull request by nbusseneau**
This PR aims to introduce transparent support for installing Cilium on clusters using the AL2023 AWS node image family.

Please see per commit for full details :)

### CODE REVIEW

File: cilium-cli/install/helm.go
Comment (auriaave) [inline cilium-cli/install/helm.go:70]: Do we know the interface naming pattern on all the image families listed (except custom)? Are all of them following eth0, except AL2023? Is it possible to retrieve the interface name directly from the host rather than infer it by the image family used?

The reason I ask is in the case where custom images based on AL2023 are used, or in the future other images follow the same convention.

File: cilium-cli/install/install.go
Comment (auriaave) [inline cilium-cli/install/install.go:234]: This part has only been moved to the aws.go file, correct? No changes impacting the above retrieval of the node image familly.

**Review by auriaave — COMMENTED** [review]
Possibly add unit tests for aws.go - otherwise, minor comments. Thanks.

File: cilium-cli/install/helm.go
Comment (nbusseneau) [reply cilium-cli/install/helm.go:70]: Neither the AWS CLI nor eksctl seem to expose this information, so it indeed would have to be retrieved from the host directly. There's the `ExecInPod` facility that could probably be used for that -- we can execute `ip address` commands and parse the output, just need a pod to leverage -- that would probably be the annoying part, since we are executing this **before** Cilium is installed.

File: cilium-cli/install/install.go
Comment (nbusseneau) [reply cilium-cli/install/install.go:234]: Correct -- I have split it in a separate commit for easier review of the move change.

**Comment by nbusseneau** [comment]
Moving from draft to ready for review to get initial feedback -- but this is not yet quite ready to merge, still running some CI tests as per the temporary commit.

File: cilium-cli/install/helm.go
Comment (auriaave) [reply cilium-cli/install/helm.go:70]: we can leave this as a future enhancement. Thanks for checking.

File: cilium-cli/install/helm.go
Comment (gandro) [inline cilium-cli/install/helm.go]: How certain are we that `ens5` is the correct interface on all nodes? Should we use a prefix match such as `ens+` here? I believe even the previous `eth0` probably should have been `eth+` to be honest, i.e. pod-to-pod traffic on secondary interfaces probably should also be subject to masquerading. 

Unfortunately our docs are a bit conflicting here. The ENI docs do recommend `eth+`, where as the Helm getting started guide just sets `eth0`:

https://docs.cilium.io/en/v1.16/network/concepts/routing/#id6
https://docs.cilium.io/en/latest/installation/k8s-install-helm/#k8s-install-helm

**Review by gandro — COMMENTED** [review]
Some inline question on the interface name

File: cilium-cli/install/aws.go
Comment (viktor-kurchenko) [inline cilium-cli/install/aws.go]: I think `eksctl get nodegroup` command will return output as JSON array, isn't it?
If yes then `json.Unmarshal` into a struct will fail.

**Comment by viktor-kurchenko** [comment]
I've pushed two commits:
1. Use K8S ListNodes API to extract OS image instead of `eksctl` tool: https://github.com/cilium/cilium/pull/35529/commits/47c8f339aba4476d52a28397180396dc83bc276a
2. Egress interfaces selector improvement: https://github.com/cilium/cilium/pull/35529/commits/0df6ea09dd5073fdb600ffe9aac0c0d73cab7ba3 (suggested by @gandro)

I can't run EKS conformance tests for the PR due to [the changes](https://github.com/cilium/cilium/commit/3a48de822bcf225cbb98456b2deca8eee9a8c06c) in the main branch.
@nbusseneau any ideas how to make CI green before merge?

I've done testing on the separate [branch](https://github.com/cilium/cilium/tree/vk/pr/al2023/test): 
- EKS 1.29: https://github.com/cilium/cilium/actions/runs/11914812464/job/33203638097
- EKS 1.30: https://github.com/cilium/cilium/actions/runs/11914812464/job/33203638539

Also, I've checked locally that it's possible to override `egressMasqueradeInterfaces` value during Cilium installation using input param: `--helm-set egressMasqueradeInterfaces=eth0`.

@gandro @auriaave @brlbil please review.

File: .github/actions/aws-cni/k8s-versions.yaml
Comment (auriaave) [inline .github/actions/aws-cni/k8s-versions.yaml]: The current file supports up to EKS 1.31 - be careful not to regress 🙏

File: .github/actions/eks/k8s-versions.yaml
Comment (auriaave) [inline .github/actions/eks/k8s-versions.yaml]: Same about https://github.com/cilium/cilium/blob/main/.github/actions/eks/k8s-versions.yaml

File: cilium-cli/install/helm.go
Comment (auriaave) [reply cilium-cli/install/helm.go:70]: This suggestion can be closed. Not pursuing it for now.

File: cilium-cli/install/aws.go
Comment (auriaave) [inline cilium-cli/install/aws.go]: Can we possibly add the 2 values that don't match? The "ami" one and the one from nodes[i]? It will contribute to add more info while troubleshooting the problem.

**Review by auriaave — APPROVED** [review]
Added a suggestion to include the mismatching OS images found in the error message. Not a blocking issue. Thanks.

File: .github/actions/aws-cni/k8s-versions.yaml
Comment (viktor-kurchenko) [reply .github/actions/aws-cni/k8s-versions.yaml]: Yeah, it's a `temp commit DO NOT MERGE` [commit](https://github.com/cilium/cilium/pull/35529/commits/d4d21e43e2cd18688f14ce56931f66ae88595ef1).
I'll remove it before merge.

File: .github/actions/eks/k8s-versions.yaml
Comment (viktor-kurchenko) [reply .github/actions/eks/k8s-versions.yaml]: Yeah, it's a `temp commit DO NOT MERGE` [commit](https://github.com/cilium/cilium/pull/35529/commits/d4d21e43e2cd18688f14ce56931f66ae88595ef1).
I'll remove it before merge.

File: cilium-cli/install/aws.go
Comment (viktor-kurchenko) [reply cilium-cli/install/aws.go]: Done, thanks!

**Review by gandro — APPROVED** [review]

**Review by learnitall — APPROVED** [review]
Docs look good to go! :rocket:

**Review by derailed — APPROVED** [review]
@nbusseneau Nice work!

File: .github/actions/eks/k8s-versions.yaml
Comment (viktor-kurchenko) [reply .github/actions/eks/k8s-versions.yaml]: Removed.

File: .github/actions/aws-cni/k8s-versions.yaml
Comment (viktor-kurchenko) [reply .github/actions/aws-cni/k8s-versions.yaml]: Removed.

**Review by viktor-kurchenko — APPROVED** [review]

**Comment by viktor-kurchenko** [comment]
I'm closing this PR in favor of https://github.com/cilium/cilium/pull/36076 because I don't know how to fix EKS workflows here.

**Pull request closed without merging**
